### PR TITLE
etcdserver: handle malformed basic auth

### LINKED ISSUE #3280: Some defect while using basic auth function
While using basic auth feature, I found that if a path has been granted to a role
I give the wrong auth header, it can still get related key. 
This behavior is different from etcdctl
Here is the my test step:

A. start etcd
  1) start etcd service on Ubuntu 14.04
	./etcd
B. user/role/path auth
1) add root user
	./etcdctl user add root
2) enable auth
        ./etcdctl auth enable
3) add new role
	./etcdctl -u root:1111 role add 0813
4) add new user
	./etcdctl -u root:1111 user add test
5) grant user to a role
	./etcdctl -u root:1111 user grant test -roles 0813
6)  Grant role to the path
	./etcdctl -u root:1111 role grant 0813 -path '/rootpath/*' -readwrite

C.get/set value

1）get/set key at rootpath
	./etcdctl -u test2:1111 get /rootpath/1
2) enable cache using url pattern
	curl -L http://127.0.0.1:2379/v2/auth/enable -XGET
3) get value without providing any user auth
	curl -L http://127.0.0.1:2379/v2/keys/rootpath/a -XGET
	{"action":"get","node":{"key":"/rootpath/a","value":"this is awesome","modifiedIndex":32,"createdIndex":32}}
**4) get value with auth header (right username and passward)**
curl -H "Authorization: Basic dGVzdDI6MTExMQ==" -L http://127.0.0.1:2379/v2/keys/rootpath/a -XGET
{"action":"get","node":{"key":"/rootpath/a","value":"this is awesome","modifiedIndex":32,"createdIndex":32}}
**5)  get value with auth header (wrong encode string), still get result**
curl -H "Authorization: Basic dGVzdDI6MTExMQ==xxxx" -L http://127.0.0.1:2379/v2/keys/rootpath/a -XGET
{"action":"get","node":{"key":"/rootpath/a","value":"this is awesome","modifiedIndex":32,"createdIndex":32}}
**6) if using etcdctl which giving random username and password, request will be denied**
./etcdctl -u 23232:1111 get /rootpath/1 
Error:  0: Insufficient credentials () [0]

### PR DESCRIPTION
**Pull request by MSamman**
return insufficient credentials if basic auth header is malformed

Fixes #3280

### CODE REVIEW

**Comment by xiang90** [comment]
LGTM. /cc @yichengq

File: etcdserver/etcdhttp/client_auth_test.go [hunk 1]
```diff
@@ -238,6 +238,50 @@ func TestAuthFlow(t *testing.T) {
...
+				enabled: true,
+				user: &auth.User{
+					User:     "root",
+					Password: goodPassword,
+					Roles:    []string{"root"},
+				},
+				roles: map[string]*auth.Role{
+					"root": {
+						Role: "root",
+					},
```
Comment (yichengq) [inline etcdserver/etcdhttp/client_auth_test.go]: I think you need to give guest role here for it to check whether it skips check on guest role.

Comment (MSamman) [inline etcdserver/etcdhttp/client_auth_test.go]: not sure I understand. Use guest role for the bad password test case or use guest role for malformed basic auth encoding test case?

Comment (yichengq) [inline etcdserver/etcdhttp/client_auth_test.go]: For  `use guest role for malformed basic auth encoding test case` case, it should fail the test before your change, and pass the test after your change.

I tested it myself, and the test could still pass when I change back the code in client_auth.go. One reason is that `req := mustJSONRequest(t, "DELETE", "enable", "")` is not a request on key space.

Comment (MSamman) [inline etcdserver/etcdhttp/client_auth_test.go]: OK. I updated with the correct test.

**Comment by yichengq** [comment]
LGTM. thanks for the contribution!

**Pull request merged**